ncclient/transport/tls.py

```diff
@@ -0,0 +1,145 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+import socket
+import sys
+import threading
+from ssl import AF_INET, CERT_REQUIRED, PROTOCOL_TLSv1_2, SOCK_STREAM, \
+    SSLContext, SSLError
+
+if sys.version < '3':
+    from six import StringIO
+else:
+    from io import BytesIO as StringIO
+
+from ncclient.capabilities import Capabilities
+from ncclient.logging_ import SessionLoggerAdapter
+from ncclient.transport.errors import TLSError
+from ncclient.transport.session import Session
+from ncclient.transport.parser import DefaultXMLParser
+
+logger = logging.getLogger("ncclient.transport.tls")
+
+DEFAULT_TLS_NETCONF_PORT = 6513
+DEFAULT_TLS_TIMEOUT = 120
+
+BUF_SIZE = 4096
+
+
+class TLSSession(Session):
+
+    def __init__(self, device_handler):
+        capabilities = Capabilities(device_handler.get_capabilities())
+        Session.__init__(self, capabilities)
+        self._host = None
+        self._connected = False
+        self._socket = None
+        self._buffer = StringIO()
+        self._device_handler = device_handler
+        self._message_list = []
+        self._closing = threading.Event()
+        self.parser = DefaultXMLParser(self)
+        self.logger = SessionLoggerAdapter(logger, {'session': self})
+
+    def _dispatch_message(self, raw):
+        self.logger.info("Received message from host")
+        self.logger.debug("Received:\n%s", raw)
+        return super(TLSSession, self)._dispatch_message(raw)
+
+    def close(self):
+        self._closing.set()
+        self._socket.close()
+        self._connected = False
+
+    def connect(self, host=None, port=DEFAULT_TLS_NETCONF_PORT,
+                keyfile=None, certfile=None, ca_certs=None,
+                timeout=DEFAULT_TLS_TIMEOUT):
+        """Establish NETCONF session via TLS.
+
+        :param host: Hostname or IP address to connect to.
+        :param port: Port number which would be used for connection, the
+            default value is 6513 (`DEFAULT_TLS_NETCONF_PORT`).
+        :param certfile: Points to file in PEM format with client's certificate
+            and optionally with client's private key.
+        :param keyfile: Points to file with client's private key in PEM format,
+            if the latter is not in certfile.
+        :param ca_certs: Points to file containing a list of certificates,
+            which should be used to verify server certificate.
+        :param timeout: Specifies the connection timeout, defaults to
+            120 seconds (`DEFAULT_TLS_TIMEOUT`).
+
+        :raise TLSError if the connection can not be established.
+        """
+        if host is None:
+            raise TLSError('Missing host')
+        if certfile is None:
+            raise TLSError('Missing client certificate file')
+
+        ssl_context = SSLContext(PROTOCOL_TLSv1_2)
```

Comment:
Hi, it seems like the `PROTOCOL_TLSv*` are getting deprecated in 3.6.
The docs suggests to use the `PROTOCOL_TLS_CLIENT` flag.

[Link to ssl.PROTOCOL_TLS_CLIENT doc](https://docs.python.org/3/library/ssl.html#ssl.PROTOCOL_TLS_CLIENT).

Reply:
Thanks for the tip! I'll address this issue in the next iteration of the PR.

Reply:
Any chance you'll able to address those changes any time soon?

I would like to help you, but I'm not sure if I can push to (or somehow edit) this pull-request as I'm not maintainer of this repository.

Reply:
I was hoping to collect all the feedback before making further code changes. @einarnn, could you have a look at the PR?

Reply:
@ivandkhn , I've started reviewing!

Reply:
@nthe, I've updated handling of TLS protocol version as you've suggested with a new `protocol` argument.